### DISCUSSION TITLE
Generating Custom Feedback & Custom Marks Per Questions

### DISCUSSION THREAD

**Discussion by HxTrinity** [jupyter/nbgrader#1847] [Q&A]
### I saw someones nbgrader doing this and I am wondering how I can generate a custom feedback form in the forms that looks like this

```
Question 1 (repeat_int_str)
Test 0 PASS
    input:      0, 0
    output:     ''
    expected:   ''

Test 1 PASS
    input:      0, 1
    output:     '0'
    expected:   '0'

Test 2 >>FAIL<<
    input:      0, 2
    output:     '000'
    expected:   '00'
    
 
Question 2 (biggest_letter_distance) - 2 Marks Each
Test 0 PASS
    input:      'Hello'
    output:     10
    expected:   10

Test 1 >>FAIL<<
    input:      'Za'
    output:     26
    expected:   25

Test 2 PASS
    input:      'AAAAaaaaAAAAAAAAaaaaaAA'
    output:     0
    expected:   0
    
Test 3 >>EXCEPTION<<
    input:      ''
    output:     IndexError('string index out of range')
    expected:   ''

 
 ```
 currently I'm doing the most basic autograde test block which is just
 
 ```python
assert somefunc(5) == [0,1,2,3,4]
assert  somefunc(7) == [0,1,2,3,4,5,6]
assert somefunc(2) == [0,1]
#etc...
 ```
 
However the issue with this is that it doesn't generate the output above and also it can't give custom marks per question. It will just stop at that specific test case when the student is wrong and the other test cases don't get run and award 0 marks.

**Comment by ranjiGT** [comment]
Hi @HxTrinity ,

what you’re seeing there isn’t the default behavior of simple `assert`-based tests—those stop at the first failure, as you noticed. The richer feedback (multiple tests, PASS/FAIL, inputs/outputs, exceptions) usually comes from **custom test code plus a custom feedback template**.

There are two parts to reproducing this:

**1) Run multiple tests without stopping on failure**
Instead of plain `assert`, you wrap your tests in a loop and catch exceptions so every case runs. For example:

```python
tests = [
    {"input": (0, 0), "expected": ""},
    {"input": (0, 1), "expected": "0"},
    {"input": (0, 2), "expected": "00"},
]

results = []

for i, t in enumerate(tests):
    try:
        out = repeat_int_str(*t["input"])
        if out == t["expected"]:
            results.append((i, "PASS", t["input"], out, t["expected"]))
        else:
            results.append((i, "FAIL", t["input"], out, t["expected"]))
    except Exception as e:
        results.append((i, "EXCEPTION", t["input"], repr(e), t["expected"]))

for r in results:
    print(f"Test {r[0]} {r[1]}")
    print(f"  input:    {r[2]}")
    print(f"  output:   {r[3]}")
    print(f"  expected: {r[4]}\n")
```

This way all tests execute and you control the output format.

**2) Custom feedback / grading in nbgrader**
nbgrader itself doesn’t natively format output like your example—you typically:

* either **print structured feedback** (as above), which shows up in the feedback notebook, or
* define a **custom Jinja2 feedback template** to control how results are rendered.

For per-question scoring, you can:

* assign points at the *cell level* (`points` in the nbgrader metadata), and
* optionally compute partial credit inside the test cell (e.g. number of passed tests) and raise an assertion only at the end if needed.

A common pattern is:

* count passed tests,
* compute a score,
* print detailed feedback,
* optionally `assert passed == total` if you want strict grading.

If you want something closer to unit-test style, you can also look into using `pytest` (some setups integrate it with nbgrader), which gives richer reporting out of the box.

In short: the behavior you saw is achieved by **replacing simple asserts with a small test harness + (optionally) custom templates**, not by a built-in nbgrader feature.

Hope that helps!
